### ISSUE TITLE
content_hash saving

### ISSUE THREAD

**Issue by us91** [jgunthorpe/cloud_mdir_sync#2]
I could not get the program to save the content_hash files ch-.* . As a result, the messages need to be downloaded every time the program is restarted. I couldn't find where msgdb.close()  or msgdb._save_content_hashes() is called.

Perhaps this has to do with how I exit the program. I couldn't find a graceful way to force the program to quit. I tried the keyboard Ctrl-C, and also sending signal TERM to the process. Neither will lead to the saving of the content hashes.

Can the content hashes be saved when the program exits, either gracefully or being forced to exit through signals?

**Comment by jgunthorpe** [comment]
ctrl-c triggers an exception and part of the exception handling already saves all the state. It is in main.py as part of the 'with .. as msgdb'

Possibly you are hitting some exception while doing the save that is being lost

**Comment by us91** [comment]
This is indeed the case. Thank you for explanation. My earlier problem was the missing of keyring module. After installing a valid keyring, saving works as expected.

It would be nice, though, if the state is saved if the process is killed through a signal (e.g,. TERM or INT). This will allow the main process to be put in the background. However, this is only a minor feature request.

**Comment by jgunthorpe** [comment]
Can you elaborate on the keyring module thing, do you think there is some logging that is missing?

SIGINT triggers the same behavior as ctrl-c, so it should already be done

**Comment by us91** [comment]
In the beginning, the content_hash was not saving. So every time cloud-mdir-sync is started, it reloads all messages. I have been using the program this way for several months, thinking this is the default behavior. Then I read some source code, thinking if I install keyring, cache will be saved. So I installed gnome-keyring (I am on Archlinux) as well as the python-keyring package (which contains the keyring module). I was prompted to enter some password for gnome-keyring, and I noticed that content_hash files are being saved.

Now at your prompt, I removed gnome-keyring, the program no longer runs, complaining missing keyring backend. I think it is wanting to use the authenticator stored in state cache. If I remove all the state cache (ch-*) files, then the program can run, like in the very beginning, but no state is saved:

`raise NoKeyringError(msg)`

In retrospect, I think the keyring module is irrelevant, as the cloud-mdir-sync script automatically set up the environment venv with the needed modules. So my problem was due to only the missing gnome-keyring backend.

I just verified that sending SIGINT to the process triggers the same `KeyboardInterrupt` and results in state saving. I incorrectly mentioned sending TERM and INT, but in fact I tried only TERM, QUIT, and KILL signals.

I think I am all set.

**Thank you so much!**

I will close the issue - as all problems related to this issue has been resolved.
